### PR TITLE
oop-static_modifier-hard-q5

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
Please see my feedback on other questions in this category and apply it to this PR.

File: src/oop/static_modifier/hard/q5/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+- Question Type: Java question
+
+## Carpet Class
+
+Your `Carpet` class must have the following requirements:
+
+- Attributes:
+    - A _static_ `double` called **materialAreaAvailable**. This represents the total number of wood beans available for
+      making `Bed` objects, and should be set initially to 300 without having to call the constructor. Note that it
+      should not be visible from other classes.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/QuestionText.md:12]:
```suggestion
    - A _static_ `double` called **materialAreaAvailable**. This represents the total number of carpet squares available for
      making `Carpet` objects, and should be set initially to 300 without having to call the constructor. Note that it
      should not be visible from other classes.
```

File: src/oop/static_modifier/hard/q5/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# oop-static_modifier-hard-q5
+
+- Question Type: Java question
+
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/QuestionText.md:4]: Carpet squares are easy to order but the colour of different batches can be slightly different and so cannot be used for the same carpet. Write a Java class called `Carpet` to keep track of the carpet squares of the most recent order using the requirements below. Then write a class called `TestCoverage` to test `Carpet`.

File: src/oop/static_modifier/hard/q5/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      **materialAreaAvailable**, meaning that the flooring cannot be made, or subtract the **houseArea** value from
+      the **materialAreaAvailable** and return true.
+
+## TestCoverage Class
+
+Your `TestCoverage` class must have the following items:
+
+- Main method:
+    - A `main` method which initializes 5 `Carpet` objects with **houseArea** of each of them being in order: 126, 133,
+      81, 100, 90.25, 156.25 then go through in order to floor them using `canCompleteFlooring` method. Finally, print
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/QuestionText.md:32]:
```suggestion
      81, 100, 90.25, 156.25. Then go through them in the same order to floor them using the `canCompleteFlooring` method. Finally, print
```

File: src/oop/static_modifier/hard/q5/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      the number of houses that can be floored in this format:
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/QuestionText.md:33]:
```suggestion
      how many of the houses can be floored in this format:
```

File: src/oop/static_modifier/hard/q5/QuestionText.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      "There is enough material for x houses" where x is the number of floorings made.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/QuestionText.md:34]:
```suggestion
      "There is enough material for the first x houses" where x is the number of floorings made.
```

Comment (SamamthaH) [reply src/oop/static_modifier/hard/q5/QuestionText.md:34]: update the other strings as well

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    public void carpetHasRequiredConstructor() {
+        String missingConstructorMessage = "The %s class is missing a required constructor. Make sure that it is named correctly and has the correct parameters.";
+        String wrongAccessModifier = "The %s class constructor has the wrong visibility modifier. Make sure that it is visible from all other classes.";
+        Class<?>[] constructorArgs = new Class[]{double.class};
+        assertTrue(testClass.hasConstructor(constructorArgs), String.format(missingConstructorMessage, className));
+        assertTrue(testClass.hasModifier(constructorArgs, "public"), String.format(wrongAccessModifier, className));
+    }
+
+    @Test
+    @Order(3)
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:67]: make this the first test

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                Arguments.of(4.5, 3.2),
+                Arguments.of(5.7, 6.9),
+                Arguments.of(6.1, 5.8),
+                Arguments.of(7.3, 2.5),
+                Arguments.of(8.2, 10.6),
+                Arguments.of(5.9, 4.4),
+                Arguments.of(4.8, 9.1),
+                Arguments.of(6.6, 1.7),
+                Arguments.of(5.5, 8.3),
+                Arguments.of(7.4, 0.5)
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:85]: test a 0 case and a case with larger numbers

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependency1")
+    @Order(4)
+    public void carpetConstructorInitializesValuesCorrectly(double value1) throws Throwable {
+        String wrongValueMessage = "The %s constructor did not initialize the %s attribute to the correct value based on the parameters passed to the constructor.";
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:94]:
```suggestion
        String wrongValueMessage = "The %s constructor did not initialize the %s attribute to the correct value based on the parameter passed to the constructor.";
```

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        Object[][] constructorArgs = {
+                {value1, double.class}
+        };
+        Object checkupInstance = testClass.createInstance(constructorArgs);
+        _assertEquals(value1, testClass.getFieldValue(checkupInstance, attributeName), String.format(wrongValueMessage, className, attributeName));
+    }
+
+    @Test
+    @Order(5)
+    public void canCompleteFlooringIsDefinedCorrectly() {
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:104]: this test is no longer needed

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    @Test
+    @Tag("dependent1")
+    @Order(7)
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral("There is enough material for 2 houses"),
+                new Optional(new StringLiteral(" "))
+        };
+        Object classInstance = classObject.createInstance();
+        classObject.callMethod("main", new Object[][]{{new String[0], String[].class}}, new String[]{"public"}, classInstance,
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:153]: add an error message

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        testClass.setFieldValue(null, initialBeams, staticAttributeName);
+    }
+
+    @Test
+    @Tag("dependent1")
+    @Order(7)
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral("There is enough material for 2 houses"),
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:149]: make this a variable at the top of the tests

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See Sam's review

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        Class<?>[] constructorArgs = new Class[]{double.class};
+        assertTrue(testClass.hasConstructor(constructorArgs), String.format(missingConstructorMessage, className));
+        assertTrue(testClass.hasModifier(constructorArgs, "public"), String.format(wrongAccessModifier, className));
+    }
+
+    @Test
+    @Order(1)
+    @Tag("dependency1")
+    public void staticVariableIsInitializedCorrectly() {
+        _assertEquals(initialBeams, testClass.getFieldValue(null, staticAttributeName),
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:71]: check that the variable exists before checking its value

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    @MethodSource("inputProvider")
+    @Tag("dependency1")
+    @Order(4)
+    public void carpetConstructorInitializesValuesCorrectly(double value1) throws Throwable {
+        String wrongValueMessage = "The %s constructor did not initialize the %s attribute to the correct value based on the parameter passed to the constructor.";
+        Object[][] constructorArgs = {
+                {value1, double.class}
+        };
+        Object checkupInstance = testClass.createInstance(constructorArgs);
+        _assertEquals(value1, testClass.getFieldValue(checkupInstance, attributeName), String.format(wrongValueMessage, className, attributeName));
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:102]: check that the constructor does not affect the static value

File: src/oop/static_modifier/hard/q5/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    @Order(7)
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral(String.format(printOutput, 2)),
+                new Optional(new StringLiteral(" "))
+        };
+        Object classInstance = classObject.createInstance();
+        classObject.callMethod("main", new Object[][]{{new String[0], String[].class}}, new String[]{"public"}, classInstance,
+                clauses,
+                "Your " + testClassName + " class main method does not print the correct output.");
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q5/MainTest.java:147]: would you be able to initialize a Carpet object before this method call and then compare it's values before and after the main method call to check if the student is actually modifying the static values?

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Pull request merged**